### DISCUSSION TITLE
Adding I2C and SPI flash to Design

### DISCUSSION THREAD

**Discussion by clikbait295** [ucb-bar/chipyard#1568] [Q&A]
Hello everyone, I'm a student trying to design my first ASIC, and I want to create a RISC-V microcontroller using a single Rocket core along with some RAM, external SPI flash and I2C capabilities.

I've been able to get RAM added onto the Rocket core, but I'm running into issues with adding external SPI flash and I2C. I've tried just adding it to the config file, but I'm running into errors. How would I go about doing this?

**Comment by milovanovic** [comment]
Hi, the generators for I2C and SPI peripherals are taken from [rocket-chip-blocks](https://github.com/chipsalliance/rocket-chip-blocks) which are the former [sifive-blocks](https://github.com/sifive/sifive-blocks). The peripherals are located inside: [/chipyard/generators/sifive-blocks/src/main/scala/devices](https://github.com/chipsalliance/rocket-chip-blocks/tree/main/src/main/scala/devices) so you can have a look. There are several examples how they're included, but perhaps you can start from the [appropriate documentation section](https://chipyard.readthedocs.io/en/stable/Generators/SiFive-Generators.html#peripheral-devices).

Try just substituting the GPIO for I2C for example like in [the following config fragment](https://github.com/ucb-bar/chipyard/blob/main/generators/chipyard/src/main/scala/config/fragments/PeripheralFragments.scala#L109) definition where you have to add the address. Similarly you should do for [the SPI flash](https://github.com/ucb-bar/chipyard/blob/main/generators/chipyard/src/main/scala/config/fragments/PeripheralFragments.scala#L88).

If you've already done so and you still get errors, you should either post them here or raise an issue.
